### PR TITLE
Part 1

### PR DESCRIPTION
**Pull request by benjimarshall**

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1,5 +1,25 @@
...
+
+function readSampleText(): string {
+    return fs.readFileSync(sampleTextPath, "utf8");
+}
+
+function naiveTls(input: string): number {
+    const inputArray = [...input];
+
+    let counter = 0;
+    for (let i = 0; i < inputArray.length - 2; i++) {
```
Comment (FridaTveit) [inline src/index.ts:13]: Could you try to do this without a for loop and without mutating a counter variable? 🙂

File: src/index.ts [hunk 2]
```diff
@@ -1,5 +1,25 @@
-function helloWorld(): void {
-    console.log('Hello world');
+import * as fs from "fs";
+
+const sampleTextPath = "data/SampleText.txt";
```
Comment (FridaTveit) [inline src/index.ts:3]: Minor but I would just define this inside the function where it's used 🙂

**Review by FridaTveit — COMMENTED** [review]
I've left a few comments, let me know if you have any questions about them 🙂

File: src/index.ts (hunk 1 above)
Comment (benjimarshall) [reply src/index.ts]: Yep! I was following the pseudocode given, but here's one based off an implementation for a later part I did. Added in 0d06628 🙂

File: src/index.ts (hunk 2 above)
Comment (benjimarshall) [reply src/index.ts]: Yep, done in ac54879 🙂

File: src/index.ts [hunk 3]
```diff
@@ -1,25 +1,23 @@
...
-
-            counter++;
-        }
-    }
-
-    return counter;
+    // Make list of all trigrams in input by splitting into chunks at offsets 0, 1, and 2
+    return _.chain(_.range(0, 3))
+        .flatMap(i => _.chunk(_.drop(inputArray, i), 3))
+        .map(trigramArray => trigramArray.join(""))
```
Comment (FridaTveit) [inline src/index.ts:16]: So here you're getting all substrings of length 3, starting at 0 then 1 then 2, right? Instead of converting the string to an array and then back to a string, can you just use the string? I think `str.match(/.{1,3}/g);` should do the same as `chunk`. But as this logic can be a bit hard to follow, I would in either case split this into a few functions:
1. A function for splitting the string into substrings of length 3, which takes an offset parameter.
2. A function for getting all trigrams for a string (which uses function 1.).

Then here you should be able to do something like:
```
return _(getAllTrigrams(intput)).filter(threeLettersRegExp).length;
``` 
(might even be able to use the native filter method instead of lodash) which is a bit easier to read 🙂

**Review by FridaTveit — COMMENTED** [review]
A few suggestions for simplifying, otherwise looks good! 🙂

File: src/index.ts (hunk 3 above)
Comment (benjimarshall) [reply src/index.ts:16]: Is there a way of dropping the first character or two from the string without turning it into an array first? I think that's why I did that. That bit of regex seems to be identical to `_.chunk`, so I definitely can swap it out but it still has the same problem in that it's greedy with the matches, meaning that it would only match `'abc'` and '`def`' from `'abcdef'`, missing out substrings like `'bcd'`. Is there some way of doing getting around this that I'm missing? 🙂

I totally agree that this needs to be split up into multiple functions and can use the native filter method, so I'll get to work on that 🙂

Comment (FridaTveit) [reply src/index.ts:16]: Substring should do what you want I think? 🙂

File: src/index.ts [hunk 4]
```diff
@@ -1,25 +1,23 @@
...
+        .map(_.toLower)
+        .filter(threeLettersRegExp.test.bind(threeLettersRegExp))
```
[the remark this answers is not in the document]
Comment (benjimarshall) [reply src/index.ts:18]: I think it might be, but I'm not sure. When I just call `filter` with `threeLettersRegExp.test` I get type errors, because `this` isn't set to the Regex. I'm still not 100% confident with the way JavaScript seems to handle the `this` keyword, but that seems to be one way to fix it.

Perhaps a more elegant way might be:
```.filter(s => threeLettersRegExp.test(s))```
What do you think? 🙂

File: src/index.ts (hunk 3 above)
Comment (benjimarshall) [reply src/index.ts:16]: Yes it does! Thanks for that 🙂

File: src/index.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (benjimarshall) [reply src/index.ts:18]: Or maybe even `.filter(s => /tra/.test(s))` 🙂

[the remark this answers is not in the document]
Comment (FridaTveit) [reply src/index.ts:18]: Yes, using arrow functions like that is preferred 🙂 Here are some articles I found useful when learning about `this` in javascript:
https://ultimatecourses.com/blog/everything-you-wanted-to-know-about-javascript-scope
https://ultimatecourses.com/blog/es6-arrow-functions-syntaxes-and-lexical-scoping
https://codeburst.io/the-simple-rules-to-this-in-javascript-35d97f31bde3

🙂

File: src/index.ts [hunk 5]
```diff
@@ -1,25 +1,23 @@
...
-            counter++;
-        }
-    }
-
-    return counter;
+    // Make list of all trigrams in input by splitting into chunks at offsets 0, 1, and 2
+    return _.chain(_.range(0, 3))
+        .flatMap(i => _.chunk(_.drop(inputArray, i), 3))
+        .map(trigramArray => trigramArray.join(""))
+        .map(_.toLower)
```
[the remark this answers is not in the document]
Comment (benjimarshall) [reply src/index.ts]: The previous map ended up getting removed in 3a98a85. I've removed this `toLower` conversion and am using the regex `/tra/i` instead to ensure case-insensitivity 🙂

File: src/index.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (benjimarshall) [reply src/index.ts]: Changed in 3a98a85 🙂

**Review by FridaTveit — APPROVED** [review]
Looks great 🙂

File: src/index.ts (hunk 3 above)
Comment (benjimarshall) [reply src/index.ts]: All done in 3a98a85 🙂 Sorry it took me a little while longer than expected, there was a subtle bug leading to slight undercounting that took me quite a while to find. Basically `.` doesn't match on all whitespace characters, which would lead to the chunking regex not splitting every three characters when passing over some whitespace; effectively changing the offset for that set of chunks. The fix I settled on was to add the `/s` flag, so all whitespace characters get matched, meaning the chunking would have a constant over the course of the matches. I hope I've explained that clearly! 🙂

Comment (FridaTveit) [reply src/index.ts]: That sounds like a tricky bug, good job spotting and fixing it! 🙂
